frontends/concrete-python/concrete/fhe/mlir/context.py

```diff
@@ -238,29 +243,660 @@ def operation(
+        x_minus_y_max = x_dtype.max() - y_dtype.min()
+
+        x_minus_y_range = [x_minus_y_min, x_minus_y_max]
+        x_minus_y_dtype = Integer.that_can_represent(x_minus_y_range)
+
+        maximum_input_bit_width = max(x.bit_width, y.bit_width)
+        if x_minus_y_dtype.bit_width <= maximum_input_bit_width:
+            return self.comparison_with_subtraction_trick(
+                resulting_type,
+                x,
+                y,
+                accept,
+                x_minus_y_dtype,
+            )
+
+        if x.original_bit_width != y.original_bit_width:
+            smaller = x if x.original_bit_width < y.original_bit_width else y
+            bigger = x if x.original_bit_width > y.original_bit_width else y
+
+            smaller_dtype = Integer(smaller.is_signed, bit_width=smaller.original_bit_width)
+            bigger_dtype = Integer(bigger.is_signed, bit_width=bigger.original_bit_width)
+
+            smaller_bounds = [
+                smaller_dtype.min(),
+                smaller_dtype.max(),
+            ]
+            clipped_bigger_bounds = [
+                np.clip(smaller_dtype.min() - 1, bigger_dtype.min(), bigger_dtype.max()),
+                np.clip(smaller_dtype.max() + 1, bigger_dtype.min(), bigger_dtype.max()),
+            ]
+
+            assert clipped_bigger_bounds[0] >= smaller_dtype.min() - 1
+            assert clipped_bigger_bounds[1] <= smaller_dtype.max() + 1
+
+            assert clipped_bigger_bounds[0] >= bigger_dtype.min()
+            assert clipped_bigger_bounds[1] <= bigger_dtype.max()
+
+            smaller_minus_clipped_bigger_range = [
+                smaller_bounds[0] - clipped_bigger_bounds[1],
+                smaller_bounds[1] - clipped_bigger_bounds[0],
+            ]
+            smaller_minus_clipped_bigger_dtype = Integer.that_can_represent(
+                smaller_minus_clipped_bigger_range
+            )
+
+            clipped_bigger_minus_smaller_range = [
+                clipped_bigger_bounds[0] - smaller_bounds[1],
+                clipped_bigger_bounds[1] - smaller_bounds[0],
+            ]
+            clipped_bigger_minus_smaller_dtype = Integer.that_can_represent(
+                clipped_bigger_minus_smaller_range
+            )
+
+            intermediate_bit_width = min(
+                smaller_minus_clipped_bigger_dtype.bit_width,
+                clipped_bigger_minus_smaller_dtype.bit_width,
+            )
+
+            if smaller.original_bit_width < intermediate_bit_width <= bigger.original_bit_width:
+                return self.comparison_with_clipping_trick(
+                    resulting_type,
+                    smaller,
+                    bigger,
+                    accept,
+                    smaller_minus_clipped_bigger_dtype,
+                    clipped_bigger_minus_smaller_dtype,
+                    smaller_bounds=(smaller_bounds[0], smaller_bounds[1]),
+                    smaller_is_lhs=(smaller is x),
```

Comment:
confusing, lhs and rhs are also interverted in self.comparison_with_clipping_trick
proposing;
compare_order= ... # either (smaller, bigger) or (bigger, smaller)
minus_order= ... # either (smaller, bigger) or (bigger, smaller)